scripts/utils/trading_strategy_helpers.js

```diff
@@ -85,37 +116,60 @@ const isOnlySafeOwner = async function(masterAddress, ownedAddress, artifacts) {
 
 /**
  * Queries EVM for ERC20 token details by address
- * and returns a list of detailed token information.
- * @param {SmartContract} exchange BatchExchange, contract, or any contract implementing `tokenIdToAddressMap`
- * @param {integer[]} tokenIds list of token ids whose data is to be fetch from EVM
- * @return {TokenObject[]} list of detailed/relevant token information
+ * and returns a list of promises of detailed token information.
+ * @param {Address[]} tokenAddresses list of *unique* token addresses whose data is to be fetch from the EVM
+ * @return {Promise<TokenObject>[]} list of detailed/relevant token information
  */
-const globalTokenObjects = {}
-const fetchTokenInfo = async function(exchange, tokenIds, artifacts, debug = false) {
+const globalTokenPromisesFromAddress = {}
+const fetchTokenInfoAtAddresses = function(tokenAddresses, artifacts, debug = false) {
   const log = debug ? () => console.log.apply(arguments) : () => {}
   const ERC20 = artifacts.require("ERC20Detailed")
 
   log("Fetching token data from EVM")
-  const tokenObjects = {}
-  await Promise.all(
-    tokenIds.map(async id => {
-      if (!(id in globalTokenObjects)) {
-        const tokenAddress = await exchange.tokenIdToAddressMap(id)
+  const tokenPromises = {}
+  for (const tokenAddress of tokenAddresses) {
+    if (!(tokenAddress in globalTokenPromisesFromAddress)) {
+      globalTokenPromisesFromAddress[tokenAddress] = (async () => {
         const tokenInstance = await ERC20.at(tokenAddress)
         const [tokenSymbol, tokenDecimals] = await Promise.all([tokenInstance.symbol.call(), tokenInstance.decimals.call()])
         const tokenInfo = {
-          id: id,
           address: tokenAddress,
           symbol: tokenSymbol,
           decimals: tokenDecimals.toNumber(),
         }
-        log(`Found Token ${tokenInfo.symbol} at ID ${tokenInfo.id} with ${tokenInfo.decimals} decimals`)
-        globalTokenObjects[id] = tokenInfo
-      }
-      tokenObjects[id] = globalTokenObjects[id]
-    })
-  )
-  return tokenObjects
+        log(`Found token ${tokenInfo.symbol} at address ${tokenInfo.address} with ${tokenInfo.decimals} decimals`)
+        return tokenInfo
+      }).call()
+      tokenPromises[tokenAddress] = globalTokenPromisesFromAddress[tokenAddress]
+    }
+  }
+  return tokenPromises
```

Comment:
Is there a reason why this method isn't async and doesn't `await` all the promises in here instead of having `await tokenPromises[tokenAddress]` for each token after the function call?

Reply:
Mmh, currently none. The original idea was that I was going to fetch any token information that I might need later on at the start of the code, so I structured the function around this. But this is indeed not the case right now. I just tried to move the token fetching at the start of the code, and the running time of the withdrawal script went from about 11s to 8s! So I'm going to keep this function as it is and to change the script accordingly instead.